statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,410 @@
+        gamma_hat_l.append(gamma_hat)
+
+        tau_hat = _gen_tau_hat(X_beta, gamma_hat, pi, p, n, alpha)
+        tau_hat_l.append(tau_hat)
+
+    return beta_hat, grad, gamma_hat_l, tau_hat_l
+
+
+def _join_debiased(model_results_l, partitions, threshold=0):
+    """
+    joins the results from each run of _est_regularized_distributed
+    and returns the debiased estimate of the coefficients
+
+    Parameters
+    ----------
+    model_results_l : list
+        A list of tuples each one containing the beta_hat, grad,
+        gamma_hat and tau_hat values for each partition.
+    partitions : scalar
+        The number of partitions that the data will be split into.
+    threshold : scalar
+        The threshold at which the coefficients will be cut.
+    """
+
+    # TODO currently the way we extract p is roundabout, should be
+    # handled better but ideally not as an argument to _join_debiased
+
+    beta_hat_l = []
+    grad_l = []
+    gamma_hat_l = []
+    tau_hat_l = []
+
+    for r in model_results_l:
+        beta_hat_l.append(r[0])
+        grad_l.append(r[1])
+        gamma_hat_l.extend(r[2])
+        tau_hat_l.extend(r[3])
+
+    p = len(gamma_hat_l)
+
+    beta_mn = np.zeros(p)
+    for beta_hat in beta_hat_l:
+        beta_mn += beta_hat
+    beta_mn = beta_mn / partitions
+
+    grad_mn = np.zeros(p)
+    for grad in grad_l:
+        grad_mn += grad
+    grad_mn = -grad_mn / partitions
+
+    theta_hat = _gen_theta_hat(gamma_hat_l, tau_hat_l, p) 
+
+    beta_tilde = beta_mn + theta_hat.dot(grad_mn)
+
+    beta_tilde[np.abs(beta_tilde) < threshold] = 0
+
+    return beta_tilde
+#    return beta_tilde, beta_mn, theta_hat, grad_mn, gamma_hat_l, theta_hat.dot(grad_mn)
+
+
+def distributed_estimation(endog_generator, exog_generator, partitions,
+                           model_class=None, init_kwds=None, fit_kwds=None,
+                           estimation_method=None, estimation_kwds=None,
+                           join_method=None, join_kwds=None):
+    """
+    This functions handles a general approach to distributed estimation,
+    the user is expected to provide generators for the data as well as
+    a model and methods for performing the estimation and recombining
+    the results
+
+    Parameters
+    ---------
+    endog_generator : generator
+        A generator to produce a sequence of endog arrays.
+    exog_generator : generator
+        A generator to produce a sequence of exog arrays.
+    partitions : scalar
+        The number of partitions that the data will be split into.
+    model_class : statsmodels model class
+        The model class which will be used for estimation. If None
+        this defaults to OLS.
+    init_kwds : dict-like or None
+        Keywords needed for initializing the model, in addition to
+        endog and exog.
+    fit_kwds : dict-like or None
+        Keywords needed for the model fitting.
+    estimation_method : function or None
```

Comment:
The name sounds like a string describing the method type.  I would use `estimator` or `estimator_func` if it is for a function argument.  

Also, function arguments are OK, but usually we use classes with methods rather than a lot of functional programming.